File: 322-coin-change/322-coin-change.cpp
class Solution {
public:
    int coinChange(vector<int>& coins, int amount) {
        vector<int> dp(amount+1,INT_MAX-1);
        dp[0] = 0;
        for(int i = 1; i <= coins.size(); i++)
        {
            for(int j = 1; j <= amount; j++)
            {
                if( coins[i - 1] <= j )
                    dp[j] = min(dp[j],dp[j-coins[i-1]]+1);
            }
        }
        if(dp[amount]==INT_MAX-1)
            return -1;
        return dp[amount];
    }
};

File: 322-coin-change/322-coin-change.java
class Solution {
    public int coinChange(int[] coins, int amount) {
        if (amount < 1) return 0;
        
        return calculate(coins, amount, new int[amount]);
    }
    public int calculate(int[] coins, int amount, int count[]) {
        
        if(amount == 0) return 0;
        if(amount < 0) return -1;
        if(count[amount-1] != 0) return count[amount-1];
        int min = Integer.MAX_VALUE;
        for (int coin: coins) {
            int rem = calculate(coins, amount-coin, count);
            if(rem >= 0 && rem < min) {
                min = rem+1;
            }
        }
        return count[amount-1] = (min == Integer.MAX_VALUE) ? -1:min;
    }
    
}